fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,573 @@
+                            )
+                        })?;
+                    if !self.ram_timestamps.contains_key(&key) {
+                        disk_newer = true;
+                        log::debug!(
+                            "Disk newer: file {} exists on disk but not in RAM",
+                            path.display()
+                        );
+                        break;
+                    }
+                }
+            }
+        }
+
+        let status = match (ram_newer, disk_newer) {
+            (false, false) => SyncStatus::InSync,
+            (true, false) => SyncStatus::StorageStale,
+            (false, true) => SyncStatus::MappingStale,
+            (true, true) => SyncStatus::Diverge,
+        };
+
+        log::info!("{} sync status is {}", self.label, status);
+        Ok(status)
+    }
+    /// Sync the in-memory storage with the storage on disk
+    fn sync(&mut self) -> Result<()> {
+        match self.sync_status()? {
+            SyncStatus::InSync => Ok(()),
+            SyncStatus::MappingStale => self.read_fs().map(|_| ()),
+            SyncStatus::StorageStale => self.write_fs().map(|_| ()),
+            SyncStatus::Diverge => {
+                let data = self.load_fs_data()?;
+                self.merge_from(&data)?;
+                self.write_fs()?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Read the data from folder storage
+    fn read_fs(&mut self) -> Result<&BTreeMap<K, V>> {
+        let data = self.load_fs_data()?;
+        self.data = data;
+        Ok(&self.data.entries)
+    }
+
+    /// Get a value from the internal mapping
+    fn get(&self, id: &K) -> Option<&V> {
+        self.data.entries.get(id)
+    }
+
+    /// Write the data to folder
+    ///
+    /// Update the modified timestamp in file metadata to avoid OS timing issues
+    /// https://github.com/ARK-Builders/ark-rust/pull/63#issuecomment-2163882227
+    fn write_fs(&mut self) -> Result<()> {
+        fs::create_dir_all(&self.path)?;
+
+        for (key, value) in &self.data.entries {
+            let file_path = self.path.join(format!("{}.bin", key));
+            let encoded: Vec<u8> = bincode::serialize(value).map_err(|e| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    format!("Failed to serialize value: {}", e),
+                )
+            })?;
+
+            let mut file = File::create(&file_path)?;
+            file.write_all(&encoded)?;
+            file.flush()?;
+
+            let new_timestamp = SystemTime::now();
+            file.set_modified(new_timestamp)?;
+            file.sync_all()?;
+
+            self.disk_timestamps
+                .insert(key.clone(), new_timestamp);
+            self.ram_timestamps
+                .insert(key.clone(), new_timestamp);
+        }
+
+        // Remove files for keys that no longer exist
+        self.remove_files_not_in_ram();
```

Comment:
Commenting out this line doesn’t break the code, and all tests pass as usual. 
Could you explain its purpose? Also, can we add tests to cover this line?

Reply:
Now that I think of it deleting a file is a fast operation (compared to writing to it). It removes the file entry from the directory inode. We could potentially implement two delete functions -
* delete - the default operations delete entries from the mapping and the file on disk
* bulk_delete - the default operations